### ISSUE TITLE
Extend API to support AWS AppSync with custom auth provider

### ISSUE THREAD

**Issue by pokryfka** [aws-amplify/amplify-swift#3210]
### Is your feature request related to a problem? Please describe.

[AWS AppSync SDK](https://github.com/awslabs/aws-mobile-appsync-sdk-ios) is now in Maintenance and users are recommended to switch to Amplify library for Swift.
However `amplify-swift` API lacks functionality (configuration) of `AWS AppSync SDK`.

Consider an existing AppSync endpoint (NOT managed by an Amplify project).
In AWS AppSync SDK the configuration requires providing of the endpoint URL and an implementation of auth provider, example:

```swift
let client = try AWSAppSyncClient(url: url, authProvider: authProvider)
```

1.

`amplify-swift` does not provide type safe interface to configure the endpoint;
the only way is via a JSON file which is:
- not type safe and
- inconvenient when using different configurations (for example for testing or QA).

```swift
/// API Category Configuration
public struct APICategoryConfiguration: CategoryConfiguration {

    /// Plugin keys to plugin configuration
    public let plugins: [String: JSONValue]

    /// Initializer for API configuration
    ///
    /// - Parameter plugins: plugin configuration map
    public init(plugins: [String: JSONValue] = [:]) {
        self.plugins = plugins
    }
}
```

2.

The [JSON configuration](https://docs.amplify.aws/lib/graphqlapi/existing-resources/q/platform/ios/) is limited to a few predefined authorization modes and does not support explicitly setting of the access token;
Note that is possible in `AWS AppSync SDK` by providing an implementation of `AWSCognitoUserPoolsAuthProviderAsync`.

```swift
public protocol AWSCognitoUserPoolsAuthProviderAsync: AWSCognitoUserPoolsAuthProvider {
    func getLatestAuthToken(_ callback: @escaping (String?, Error?) -> Void)
}
```

(Its a critical functionality in my projects which use custom authentication endpoint.)

### Describe the solution you'd like

Please:

1. extend `amplify-swift` API client to allow custom implementation AppSync auth provider
2. extend `amplify-swift` API client to support type safe explicit configuration of AppSync endpoints
3. allow explicit configuration of subscription endpoints (currently its hardcoded), see https://github.com/awslabs/aws-mobile-appsync-sdk-ios/issues/506

### Describe alternatives you've considered

Looking into 3rd party GraphQL clients for Swift.
However I would prefer to keep using a library backed up by AWS.

**Comment by phantumcode** [comment]
@pokryfka Thanks for submitting the issue.

Have you look at using the Amplify Auth category to support your authentication use cases? https://docs.amplify.aws/lib/auth/getting-started/q/platform/ios/

Please also refer to the Amplify API category documentation for additional details on setting up and using API: https://docs.amplify.aws/lib/graphqlapi/getting-started/q/platform/ios/#initialize-amplify-api

There is also several authorization modes supported in the API category.  Documentation can be found here: https://docs.amplify.aws/lib/graphqlapi/authz/q/platform/ios/#oidc

The Amplify API category can be configured using the Amplify CLI.  This would manage and generate the `amplifyconfiguration.json` file which automatically used to configure the API plugin at Amplify initialization and configuration time.

**Comment by pokryfka** [comment]
@phantumcode Thank you for the references.

I have not tested it but https://docs.amplify.aws/lib/graphqlapi/authz/q/platform/ios/#oidc seems to allow implementation of custom auth provider for AppSync.

However as of now it is inconvenient and not type safe to configure that.

Please correct me if I am wrong but it does not seem possible at all to use Amplify CLI **without** a Amplify project in AWS?
I just want to use existing AppSync endpoints - which I deploy via `serverless` - and I dont need AWS Amplify project for that.

Is it possible to extend API of the library so that Amplify CLI and the JSON configuration file are optional?
Otherwise `amplify-swift` is not really a replacement for `AWS AppSync SDK`

**Comment by lawmicha** [comment]
Hi @pokryfka, thank you for the details on describing the solutions you'd like. I agree a type-safe way to provide the configuration will make it less coupled with the Amplify CLI's `amplifyconfiguration.json` file. In the current architecture, the entry point in configuring the plugins is two steps:

1. Adding the plugin to Amplify, ie. `Amplify.add(plugin: AWSAPIPlugin())`
2. Calling `Amplify.configure()` or `Amplify.configure(config)`

If calling `Amplify.configure()`, this will use `amplifyconfiguration.json`. If calling `Amplify.configure(config)`, `config` is of type [AmplifyConfiguration](https://github.com/aws-amplify/amplify-swift/blob/main/Amplify/Core/Configuration/AmplifyConfiguration.swift#L15)

I can see that if we go the route of using an instance of AmplifyConfiguration, then we would have to create the configuration for every plugin in-memory. We can't have, for example, the auth configuration in `amplifyconfiguration.json` while having the AWSAPIPlugin's configuration in-memory. It is either the entire configuration from `amplifyconfiguration.json` or the entire config from the AmplifyConfiguration object. I assume this may not be reasonable for you since you will most likely have the AWSCognitoAuthPlugin and that configuration is quite large, generated in `amplifyconfiguration.json`, but feel free to see if using `AmplifyConfiguration` gets your code in a place that works for you.

What if we exposed a public configuration struct from AWSAPIPlugin and extend its initializer to take it in, and implement that it overrides conflicting sections in the the `amplifyconfiguration.json` file/AmplifyConfiguration object?

The DX could look something like this

```swift
// `amplifyconfiguration.json` is added to the target, and may contain other plugin configurations
// ie. storage, auth, etc.

// Create an in-memory AWSAPIPlugin configuration that will override any config
// from `amplifyconfiguration.json`
let configuration = AWSAPIPluginConfiguration(
    apis: ["apiName": .init(endpointType: .graphQL,
                            endpoint: URL(string: urlString),
                            region: "us-east-1",
                            authorizationType: .userPool)])
Amplify.add(plugin: AWSAPIPlugin(configuration: configuration))
Amplify.configure()
```

Would this developer experience solve the problem for you?

**Comment by pokryfka** [comment]
Thank you following up on that.

Personally I think having a type safe well defined configuration type, like in your example, will be very convenient.
Especially for project using AppSync without Amplify (mine scenario and all other projects which used to use AppSync SDK).
Please note, or correct me if I am wrong here, that the JSON file cannot be generated without an Amplify project. Hence I need to reverse engineer the JSON file syntax.

I think either JSON file or in memory is very reasonable.

If I understand documentation correctly, In my case, I'd need to use `OPENID_CONNECT` `authorizationType` and then add  auth provider factory:
```swift
try Amplify.add(plugin: AWSAPIPlugin(apiAuthProviderFactory: MyAPIAuthProviderFactory()))
```
I am assuming this would still work with in memory configuration.

**Comment by lawmicha** [comment]
Yes that's right, if you use AppSync without an amplify project, then you won't be able to get the `amplifyconfiguration.json` file and have to manually create this file (or create the in-memory AmplifyConfiguration object). This fields are documented here: https://docs.amplify.aws/lib/graphqlapi/existing-resources/q/platform/ios/ 

There are a few options to defining your own interceptor, the one you have above is the ODIC use case. You can also set it to `NONE` and create a URLRequestInterceptor (https://docs.amplify.aws/lib/graphqlapi/authz/q/platform/ios/#none).

How are you defining the type to decode the response from AppSync? Do you hand write these? I'm aware that Amplify CLI provides a few tools to create types to use in the app
- `amplify codegen types` produces the  `API.swift`
- `amplify codegen models` produces the swift model files from the source schema (the source schema is the one that's annotated using amplify directives like `@model`)

**Comment by pokryfka** [comment]
I tried to configure `swift-amplify` but the more I get familiar with ir the more it seems that its by design VERY different than [AWS AppSync SDK](https://github.com/awslabs/aws-mobile-appsync-sdk-ios) and cannot be used as its replacement.

I used to generate Swift code using GQL schema and GQL query/mutations/subscriptions with
`amplify codegen types`.

But it does not seem possible to use GQL in amplify, unless GQL types are decorated with proprietary decorators like `@model` or `@searchable` in which case `amplify codegen models` a lot of CRUD queries and mutations in an opinionated way.

--

TL;DR

@lawmicha I appreciate your comments but I dont think its possible to use `amplify-swift` as GraphQL (AppSync) client the way AWS AppSync SDK works; they both have been designed to work very differently (even though amplify can use AppSync under the hood)

**Comment by pokryfka** [comment]
> * `amplify codegen types` produces the  `API.swift`

^ this is the way I generate code

**Comment by lawmicha** [comment]
In Amplify CLI 12.3.0, we introduced a compatibility layer so you can re-use API.swift with Amplify.API, please take a look at https://docs.amplify.aws/lib/graphqlapi/upgrade-guide/q/platform/ios/

**Comment by pokryfka** [comment]
> In Amplify CLI 12.3.0, we introduced a compatibility layer so you can re-use API.swift with Amplify.API, please take a look at https://docs.amplify.aws/lib/graphqlapi/upgrade-guide/q/platform/ios/

I appreciate your prompt response.

I think while `AWS AppSync SDK` was an implementation of GraphQL client (based on Apollo and tested by AWS team to work with AppSync),
`amplify-swift` tries to do much much more than just being a GraphQL client.

This may have some benefits - for example, it generates cloud formation templates for AWS resources (adding `@searchable` sets up ElasticSearch stack!), or Amplify Admin out of a box.

At the same time it adds a lot of complexity and imports many dependencies (like all AWS SDK) **if** only AppSync integration is needed.

For my needs I just need a replacement of GraphQL client; I would prefer to use a library supported by AWS/AppSync teams but I think Amplify is just way too heavy.
Currently testing [swift-graphql](https://github.com/maticzav/swift-graphql) and it seems to cover all my needs.

**Comment by lawmicha** [comment]
Thank you @pokryfka for your feedback and sharing your experience with us!
